src/OpenSage.Game/Graphics/ModelInstance.cs

```diff
@@ -34,6 +32,12 @@ public sealed class ModelInstance : DisposableBase
         /// </summary>
         internal readonly bool[] BoneVisibilities;
 
+        /// <summary>
+        /// Calculated bone visibilities for the current frame. Child bones will be hidden
+        /// if their parent bones are hidden.
+        /// </summary>
+        internal readonly bool[] BoneFrameVisibilities;
```

Comment:
I can’t see anywhere that sets BoneVisibilities[i] now? What is the difference between BoneVisibilities and BoneFrameVisibilities?

Reply:
BoneVisibilities[i] is set in W3dModelDraw (line 317 -> 343) (the 'ShowSubObject' and 'HideSubObject' stuff)
So BoneVisibilites are the values set by the drawModules and BoneFrameVisibilites are the values derived for each frame from BoneVisibilites, parent visability and ModelBoneInstances[i].Visible.

Reply:
see line 174 in ModelINstance.cs